### PR TITLE
Add inlay hints support

### PR DESCRIPTION
**Pull request by Kingwl**
Looks not bad:
image

It should be merge after https://github.com/microsoft/vscode/pull/127391 released.

/cc: @orta

### CODE REVIEW

**Comment by orta** [comment]
Nice!

**Comment by Kingwl** [comment]
Upstream has merged. But I guess we need wait for the release.

**Comment by hediet** [comment]
Unfortunately, this also depends on the release of TypeScript 4.4.0.
I don't want to reference an unstable version of TypeScript.

**Comment by Kingwl** [comment]
That make sense.

File: src/tsWorker.ts [hunk 1]
```diff
@@ -415,6 +415,29 @@ export class TypeScriptWorker implements ts.LanguageServiceHost, ITypeScriptWork
...
 		this._extraLibs = extraLibs;
 	}
+
+	async provideInlayHints(fileName: string, start: number, end: number): Promise<readonly ts.InlayHint[]> {
+		if (fileNameIsLib(fileName)) {
+			return [];
+		}
+		const preferences: ts.InlayHintsOptions = {
+			includeInlayParameterNameHints: "all"
+		};
```
Comment (spahnke) [inline src/tsWorker.ts:425]: Could this be made configurable via `monaco.languages.javascriptDefaults`/`monaco.languages.typescriptDefaults`? I know we don't really have a precedent for such a deep configuration in `monaco-typescript`, right now. Speaking from my own experience, I love this feature in Visual Studio, but only for literal values where it's not obvious what the parameter is for. And especially as someone who uses the Monaco editor to teach beginner programmers, inlay hints for literal values would be awesome to have, but anything else probably overwhelming. A configuration option would be greatly appreciated 🙇‍♂️

Comment (Kingwl) [reply src/tsWorker.ts:425]: Yep sure. I suppose to add some options.

Comment (orta) [reply src/tsWorker.ts:425]: https://github.com/microsoft/monaco-typescript/commit/26d78129faf0c9a6b1b8fbf848ca5ee78bcdaea2 is an example of adding a config variable to monaco-typescript

Comment (Kingwl) [reply src/tsWorker.ts:425]: I'm not sure should I add a new option like 'inlayHintsOptions' or just add these into some existed option?

File: src/languageFeatures.ts [hunk 2]
```diff
@@ -574,9 +576,9 @@ function tagToString(tag: ts.JSDocTagInfo): string {
 	if (tag.name === 'param' && tag.text) {
 		const [paramName, ...rest] = tag.text;
 		tagLabel += `\`${paramName.text}\``;
-		if (rest.length > 0) tagLabel += ` — ${rest.map(r => r.text).join(' ')}`;
+		if (rest.length > 0) tagLabel += ` — ${rest.map((r) => r.text).join(' ')}`;
```
Comment (Kingwl) [inline src/languageFeatures.ts:579]: I just run `npm run prettier`

**Review by orta — APPROVED** [review]
Looks pretty reasonable to me 👍🏻

**Comment by Kingwl** [comment]
Let's find out what can we do with inlay hints😏

**Comment by hediet** [comment]
Please ping me once TypeScript 4.4.0 (stable) is available and I'll merge this PR and do a release!

**Comment by Kingwl** [comment]
@hediet ping😂

**Comment by hediet** [comment]
Oh sorry, I missed that!

**Comment by QuevedoIB** [comment]
After upgrading from monaco-editor 0.27.0 to 0.28.1 I get the following error

image

I'm using monaco without importing 'monaco-editor/esm/vs/basic-languages/monaco.contribution'; In order to make it work with tmLanguages for theming.

Is there any way to get rid of this error or make inlay hints work without importing basic-languages?

Thanks in advance

**Comment by hediet** [comment]
@QuevedoIB how can I reproduce that?

**Pull request merged**